pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java

```diff
@@ -0,0 +1,156 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.segment.local.io.util;
+
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import org.apache.pinot.segment.spi.memory.PinotDataBuffer;
+
+
+public class ValueReaderComparisons {
+
+  private ValueReaderComparisons() {
+  }
+
+  private static int mismatch(PinotDataBuffer dataBuffer, long startOffset, int length, ByteBuffer buffer) {
+    boolean littleEndian = dataBuffer.order() == ByteOrder.LITTLE_ENDIAN;
+    if (littleEndian) {
+      buffer.order(ByteOrder.LITTLE_ENDIAN);
+    }
+    int limit = Math.min(length, buffer.limit());
+    int loopBound = limit & ~0x7;
+    int i = 0;
+    for (; i < loopBound; i += 8) {
+      long ours = dataBuffer.getLong(startOffset + i);
+      long theirs = buffer.getLong(i);
+      if (ours != theirs) {
+        long difference = ours ^ theirs;
+        return i + ((littleEndian ? Long.numberOfTrailingZeros(difference) : Long.numberOfLeadingZeros(difference))
+            >>> 3);
+      }
+    }
+    for (; i < limit; i++) {
+      byte ours = dataBuffer.getByte(startOffset + i);
+      byte theirs = buffer.get(i);
+      if (ours != theirs) {
+        return i;
+      }
+    }
+    return -1;
+  }
+
+  static int compareUtf8Bytes(PinotDataBuffer dataBuffer, long startOffset, int length, boolean padded, byte[] bytes) {
+    // can use MethodHandles.byteArrayViewVarHandle here after dropping JDK8
+    ByteBuffer buffer = ByteBuffer.wrap(bytes);
+    int mismatchPosition = mismatch(dataBuffer, startOffset, length, buffer);
+    if (mismatchPosition == -1) {
+      if (padded && bytes.length < length) {
+        // need to figure out whether the unpadded string is longer than the parameter or not
+        if (bytes.length == 0) {
+          // just need nonzero first byte
+          return dataBuffer.getByte(startOffset) == 0 ? 0 : 1;
+        } else if (bytes[bytes.length - 1] == 0) {
+          // we can't store zero except as padding, but can have it in parameters, so check the last byte
+          return dataBuffer.getByte(startOffset + bytes.length - 1) == 0 ? -1 : 0;
+        } else {
+          // check if the stored string continues beyond the length of the parameter
+          return dataBuffer.getByte(startOffset + bytes.length) == 0 ? 0 : 1;
+        }
+      } else {
+        // then we know the length precisely or know that the parameter is at least as long as we can store
+        return length - bytes.length;
+      }
+    }
+    // we know the position of the mismatch but need to do utf8 decoding before comparison
+    // to respect collation rules
```

Comment:
Could you help add an example in the comment to help understand the need of compareUtf8()? 

Would compareUtf8() return 0? If so, is there need to compare the remaining bytes after the mismatchPosition?

Reply:
Hi @klsince it's mostly to do with comparing invalid UTF8 with non ASCII text. For example, comparing the strings 
"𠜎" (UTF8=f0a09c8e) and the garbage string which renders as "�ػ" (UTF8=efbfbdd8bb) then there is a mismatch at zero, but 
```java
Byte.compareUnsigned((byte)0xf0, (byte)0xef);
``` 
gives you -1, but 
```java
new String(BytesUtils.toByteArray("f0a09c8e"), StandardCharsets.UTF_8).compareTo(new String(BytesUtils.toByteArray("efbfbdd8bb"));
``` 
agrees with the implementation on this branch that the comparison is 10172, because f0a09c8e is a valid UTF8 sequence and efbfbdd8bb isn't.

As for whether `compareUtf8` can return zero, that's a good question. It could return zero for two invalid UTF8 sequences, so it will need more testing against `String`. Thanks for raising this.

Reply:
I just added more testing and I can't make `compareUtf8` produce zero if either the stored data or the parameter were obtained from `String` instance.